test/k8s/res_limits_test.go

```diff
@@ -0,0 +1,485 @@
+
+			adjustResource(cpu, memory, false)
+
+			deployDPs()
+
+			errCh := make(chan error)
+			ctx, cancelMonitoring := context.WithCancel(context.Background())
+			defer cancelMonitoring()
+			watchControlPlane(ctx, errCh)
+
+			dpCh := make(chan bool)
+			waitForDPs(dpCh)
+
+			var err error
+			select {
+			case dpRet := <-dpCh:
+				Logf("dpCh returned\n")
+				Expect(dpRet).To(BeTrue(), "data planes should be run and available")
+			case err = <-errCh:
+				Logf("errCh returned\n")
+			}
+			Expect(err).ToNot(HaveOccurred(), "control plane should not crash")
+		})
+
+		// half the resource, should crash without GOMEMLIMIT
+		It("should crash when deploy all services and instances with half CP resource", func() {
+			adjustResource(cpu/2, memory/2, false)
+
+			deployDPs()
+
+			errCh := make(chan error)
+			ctx, cancelMonitoring := context.WithCancel(context.Background())
+			watchControlPlane(ctx, errCh)
+
+			dpCh := make(chan bool)
+			waitForDPs(dpCh)
+
+			var err error
+			select {
+			case <-dpCh:
+				Logf("dpCh returned\n")
+			case err = <-errCh:
+				Logf("errCh returned\n")
+			}
+			cancelMonitoring()
+			Expect(err).To(HaveOccurred(), "control plane should crash with half resource")
+
+			cpPods, err := k8s.ListPodsE(cluster.GetTesting(), cluster.GetKubectlOptions(Config.KumaNamespace), metav1.ListOptions{
+				LabelSelector: fmt.Sprintf("app=%s", Config.KumaServiceName),
+			})
+			Expect(err).ToNot(HaveOccurred())
+			for _, pod := range cpPods {
+				if k8s.IsPodAvailable(&pod) {
+					continue
+				}
+
+				_, _ = fmt.Fprintf(GinkgoWriter, "Pod %s: %s\n", pod.Name, pod.Status.Phase)
+				for _, cts := range pod.Status.ContainerStatuses {
+					_, _ = fmt.Fprintf(GinkgoWriter, "Pod %s - %s: %s\n", pod.Name, cts.Name, cts.State.String())
+				}
+			}
+		})
+
+		It("should not crash when control plane has GOMEMLIMIT set when all services and instances with half CP resource", func() {
+			adjustResource(cpu/2, memory/2, true)
+
+			deployDPs()
+
+			errCh := make(chan error)
+			ctx, cancelMonitoring := context.WithCancel(context.Background())
+			watchControlPlane(ctx, errCh)
+
+			dpCh := make(chan bool)
+			waitForDPs(dpCh)
+
+			var err error
+			select {
+			case <-dpCh:
+				Logf("dpCh returned\n")
+			case err = <-errCh:
+				Logf("errCh returned\n")
+			}
+			cancelMonitoring()
+			Expect(err).ToNot(HaveOccurred(), "control plane should not crash")
+			if err != nil {
+				_, _ = fmt.Fprint(GinkgoWriter, err.Error())
+			}
+
+			dpPods, err := k8s.ListPodsE(cluster.GetTesting(), cluster.GetKubectlOptions(TestNamespace), metav1.ListOptions{})
+			Expect(err).ToNot(HaveOccurred())
+			for _, pod := range dpPods {
+				if k8s.IsPodAvailable(&pod) {
+					continue
+				}
+
+				_, _ = fmt.Fprintf(GinkgoWriter, "Pod %s: %s\n", pod.Name, pod.Status.Phase)
+				for _, cts := range pod.Status.ContainerStatuses {
+					_, _ = fmt.Fprintf(GinkgoWriter, "Pod %s - %s: %s\n", pod.Name, cts.Name, cts.State.String())
+				}
```

Comment:
Why are we printing this? Should this be assertion or what's the reason?

Reply:
I want to print dp statuses to help diagnose issues. For the case when "control plane should not crash", the DPs are also expected to be healthy. So if there is any DP unready, it should be not related to the CP.